README.md
![GPU Voxels Logo](http://gpu-voxels.org/gpu-voxels-logo.png  "GPU Voxels")

**GPU-Voxels** is a CUDA based library which allows high resolution volumetric collision detection
between animated 3D models and live pointclouds from 3D sensors of all kinds.
It is mainly developed for robotics planning and monitoring tasks.
See our videos on http://www.gpu-voxels.org/demos/ to get an impression.

## Support:
Find more information on our website http://www.gpu-voxels.org
Doxygen is available at http://www.gpu-voxels.org/doxygen/html/

If you have questions or experience any problems with the software, please post to our GitHub Bugtracker:
at https://github.com/fzi-forschungszentrum-informatik/gpu-voxels/issues

## Install dependencies:
This software is tested under 64 Bit Ubuntu Linux 14.04, 16.04 and 18.04.
Find detailed installation and linking instructions in our Doxygen.

**Core:**

- CUDA 7.5, 8.0, 9.x or 10.0
- PCL
- OpenNI
- Boost
- TinyXML (libtinyxml-dev)

**Visualizer:**

- GLEW (libglew, libglew-dev)
- GLM (libglm-dev). Probably you will have to patch one file: See https://github.com/g-truc/glm/issues/530
- OpenGL
- GLUT (freeglut3, freeglut3-dev)

**Examples:**

- ROS (Robot Operation System)
- OpenNI to interface 3D cameras
- urdfdom
- orocos
- kdl_parser

## Compiling:
        mkdir build
        cd build
        cmake ..
        make

==> This will also generate example programs and a visualization tool to see some live demos.

Doxygen files can be generated by

        make doc

## Compiling without C++11
C++11 is enabled by default. To compile without C++11 mode comment out this at the top of packages/gpu_voxels/CMakeLists.txt: 
 SET(CMAKE_CXX_STANDARD 11)
This is important if you are still using ROS Indigo and need to compile without C++11 support for compatibility reasons. There is a separate indigo branch that only differs in this line to disable C++11 during compilation.

## Known issues
- If the ROS dependency was found, but the GPU-Voxels URDF features are still unabailable, run `source /opt/ros/YOUR_ROS_DISTRO/setup.bash` before running cmake.
- Eigen 3 issues: can be fixed by cloning a more current unstable Eigen version and placing it in CMAKE_PREFIX_PATH
  + on Ubuntu 18.04 with CUDA 10.0: "math_functions.hpp not found"
  + Eigen 3.3.4 and 3.3.5 with CUDA 9.0, 9.1, 9.2: Error: class "Eigen::half" has no member "x"
  + see http://eigen.tuxfamily.org/index.php?title=Main_Page#Download
- Cuda 8.0: Code compiled with Cuda 8.0 works fine with older GPU drivers such as 375.66, but there are runtime errors with driver 384.111 and newer ("PTX JIT compilation failed"). Easy fix: use Cuda 10 with a current 410 or newer driver version. Cuda 10 is also available for Ubuntu 14.04 and 16.04.
- GLM: There is a known bug in GLM on Ubuntu 16.04 that has to be patched to allow usage of the visualizer. Apply the patch described at https://github.com/g-truc/glm/issues/530 to /usr/include/glm/detail/func_common.inl

## Maintainers:
The library is developed and maintained by:

- Christian Jülg
- Andreas Hermann

Contributors are (in temporal order):

- Sebastian Klemm
- Florian Drews
- Matthias Wagner
- Felix Mauch
- Herbert Pietrzyk
- Gabriele Bolano
- Florian Fervers
- Erik Albert

## License:
The software in this repository underlies different licenses.
Please consider the LICENSE.txt in the different package directories.
Feel free to contact us for any licensing questions.

To clarify this:

- The Build System (icmaker and according scripts) is distributed under BSD license
- GPU-Voxels is distributed under the CDDL license
- icl_core helber library is distributed under CDDL license

Please let us know if you are using the **GPU-Voxels** library,
as we are curious to find out how it enables other people's work or research.

We would be grateful if scientific publications resulting from projects
that make use of **GPU-Voxels** would cite our overview paper:
*A. Hermann, F. Drews, J. Bauer, S. Klemm, A. Roennau and R. Dillmann:
**“Unified GPU Voxel Collision Detection for Mobile Manipulation Planning”**
in Intelligent Robots and Systems (IROS 2014, Chicago, September 2014)*

Copyright (c) 2014, *FZI Forschungszentrum Informatik*, Karlsruhe, Germany
All rights reserved.

## Thanks:
The **GPU-Voxels** library was developed at the *FZI Forschungszentrum Informatik*, Karlsruhe, Germany.

The research leading to the results has received funding from... 
- the **German Federal Ministry of Education and Research (BMBF)** under grant agreement no. 01IM12006C (ISABEL).
- the **European Union** program **Horizon 2020** under grant agreement no. 680734 (HORSE).
- the **Baden-Württemberg Stiftung** through the KolRob reasearch project.



changelog.txt
Changelog of GPU-Voxels:

V 0.1 (2014-09-29):
===================
- Initial Version of GPU Voxels 
  (Please find a list of features on the Website or the according scientific papers or have a look at the example programs)
- Open Issue: Compiled only in Debug Mode

V 0.2 (2014-09-12):
===================
- Fixed compilation issued in Release Mode

V 0.3 (2015-02-26):
===================
- Made changes in the include structure to allow compilation with CUDA 6.5 (no longer trying to compile Boost-SharedMem-Headers with NVCC)
- Added datatype "PrimitiveArray" to quickly visualize helpers like thousands of spheres or cubes.
- Added InsertMetaPointCloud option to give each sub-cloud an own voxel type
- Function to sync a MetaPointCloud back to Host Memory
- Speed up in clearing VoxelMaps
- Added an optional offset when intersecting maps. So you can move one map relative to another before intersecting them.
- Added parameter to Kinect Interface to support the new PCL interface where a Cam ID has to be specified.
- More documentation in Octree kernels
- Added more operators for primitive data types
- Added Binvox as file format to generate Pointclouds
- Added Example program to demonstrate the insertion of IBEO Sensor Pointclouds
- Added missing config file for example program that intersects live Kinect data with a robot model (RobotVsEnvironment)
- Open Issues: Flickering colors in Visualization

V 0.4 (2015-03-09):
===================
- Fixed coordinate system in the Viewer
- Improved Viewer XML Config parser
- Added possibility to control visibility of SweptVolume subvolumes from a provider program
- Improved pointcloud file loader (new datatype, scaling)
- Improved example programs, added viwer configfile for them
- Added missing pointcloud filess for the examples
- Moved performance monitor to icl_core
- Some smaller fixes regarding RAD / DEG
- Removed compiler warnings

V 0.5 (2015-04-13):
===================
- Release 0.5: Critical bugfixes and reduction of warnings.
- Fixed loadbalancer bug that occurred when inserting only very little data into an octree.
- Fixed bug in extraction of Octree data.
- Fixed color gradient along Z-Axis in viewer.
- Fixed overwrite of Swept-Volume subvolumes via provider in the Viewer.
- Fixed heaps of compiler warnings (Thanks to Felix Mauch).
- Renamed MapTypes to more understandable names.
- Implemented disk-reader and writer for generic Voxelmaps.
- Simplified file loading of pointclouds.
- Moved some math functions to more general file.
- Moved some DH computation code in preparation for URDF support.
- Reimplemented function to move focus point in Viewer.
- Removed dead code and dead files.

V 0.51 (2015-04-24):
====================
- Synced readme files.

V 0.6 (2015-04-24):
===================
- Fixed crahses of octree_provider command line interface resulting from race-conditions.
- Fixed flickering in Visualizer when drawing more then one Octree. Resulted from same race-conditions as above.
- Corrected axis order in visualizer console output.
- Fixed model loading path when not using GPU_VOXELS_MODEL_PATH
- Changed argument handling in command line interface for loading models from PCD only to more general pointcloud files.

V 0.7 (2015-06-10):
===================
- Added URDF parser and interfaces to load ROS robots.
- Added URDF example program with HoLLiE robot model and URDF file.
- Unified API for DH-Parameter and URDF robots.
- Added name lookup to MetaPointClouds.
- Restructured CMake files to generate more descriptive report about missing libs.
- Fixed missing thrust include.
- Fixed color gradient in visualizer.
- Fixed DH transformations for prismatic links.

V 0.71 (2015-06-10):
====================
- Updated this changelog.

V 0.8 (2015-11-09):
===================
- Compiles with CUDA 7.5
- Added Voxellists as new Datatype
    - Huge speedup compared to Voxelmaps with sparse data.
    - Added Bit-Shifting-Collisionchecks (Thanks to Felix Mauch)
    - Added Bit-Margin for Bitvoxel Collisionchecks (Thanks to Felix Mauch)
- General:
    - Removed superficial memory allocation from Voxelmap.
    - Added script to automatize binvox mesh voxelization
    - Improved PrimitiveArrays update speed when number of primitives stays constant.
    - Removed heaps of compiler warnings (Thanks to Christian Juelg)
    - Various Bugfixes, mainly in collision Kernels
- Renamings / moved code:
    - HighLevel-API: Added "merge" function.
    - CudaMath now longer requires an instance. Now it is a namespace.
    - VoxelType ==> BitVoxelMeaning
    - Fixed standard CTORs of basic datatypes for CUDA shared memory allocation.
    - Moved voxels to gpu_voxels namespace
    - Removed seom missleading typedefs
    - Moved Voxel definitions to own subfolder.
    - Removed dead code fragments.
- Octree:
    - Adapted to new API.
    - Removed its own voxellist. Now uses more general Voxellist.
    - Removed VoxelTypeFlags and replaced them by more general bitvectors.
    - Simplified Octree selftests and triggered them by Boost-Test.
    - Dropped CUB lib. Now using CUDAs version from Thrust.
- Robot representation:
    - Enhanced robot interface (jointnames and min max angles).
    - Restructured inheritance of robot implementations.
    - Fixed install of robot headers.
- Visualizer:
    - Updated Visualizer to new API.
    - Added new function to move cam in plane.
    - Switched from Octree to more general Cubelist interface,
      as this can also be used to visualize Voxellists.
    - Visualizer now cleans Shared-Mem files.
    - Added Script to generate color gradient config files for visualizer.
    - Fixed a bug in drawing the grid
- Example programs and Testcases:
    - Updated examples to new API.
    - Added more Boost-Tests.
    - Reordered model files.
- icl_core and basic libs:
    - Added performance monitor docu
    - Some new cmake magic
    - Improved ringbuffer
    - New Doxygen version.
    - Tinyxml cmake updated.

example_how_to_link/gvl_linkage_test.cpp
#include <iostream>
using namespace std;
#include <gpu_voxels/logging/logging_gpu_voxels.h>

#include "gvl_linkage_test_lib.h"
#include <stdlib.h>

int main(int argc, char **argv)
{
    // Always initialize our logging framework as a first step, with the CLI arguments:
    icl_core::logging::initialize(argc, argv);

    std::cout << "Creating my functional object" << std::endl;
    gvlLinkageTest my_class;

    return my_class.doFancyStuff();
}


gvl_ompl_planning/gvl_ompl_planner.cpp
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK ----------------------------------------------
// This file is part of the GPU Voxels Software Library.
//
// This program is free software licensed under the CDDL
// (COMMON DEVELOPMENT AND DISTRIBUTION LICENSE Version 1.0).
// You can find a copy of this license in LICENSE.txt in the top
// directory of the source code.
//
// © Copyright 2018 FZI Forschungszentrum Informatik, Karlsruhe, Germany
//
// -- END LICENSE BLOCK ------------------------------------------------

//----------------------------------------------------------------------
/*!\file
 *
 * \author  Andreas Hermann
 * \date    2018-01-07
 *
 */
//----------------------------------------------------------------------/*
#include <iostream>
using namespace std;
#include <gpu_voxels/logging/logging_gpu_voxels.h>

#define IC_PERFORMANCE_MONITOR
#include <icl_core_performance_monitor/PerformanceMonitor.h>

#include <ompl/geometric/planners/sbl/SBL.h>
#include <ompl/geometric/planners/kpiece/LBKPIECE1.h>

#include <ompl/geometric/PathSimplifier.h>

#include "gvl_ompl_planner_helper.h"
#include <stdlib.h>

#include <memory>

namespace ob = ompl::base;
namespace og = ompl::geometric;




int main(int argc, char **argv)
{
    // Always initialize our logging framework as a first step, with the CLI arguments:
    icl_core::logging::initialize(argc, argv);

    PERF_MON_INITIALIZE(100, 1000);
    PERF_MON_ENABLE("planning");

    // construct the state space we are planning in
    auto space(std::make_shared<ob::RealVectorStateSpace>(6));
    //We then set the bounds for the R3 component of this state space:
    ob::RealVectorBounds bounds(6);
    bounds.setLow(-3.14159265);
    bounds.setHigh(3.14159265);

    bounds.setHigh(1, 0.0);

    space->setBounds(bounds);
    //Create an instance of ompl::base::SpaceInformation for the state space
    auto si(std::make_shared<ob::SpaceInformation>(space));
    //Set the state validity checker
    std::shared_ptr<GvlOmplPlannerHelper> my_class_ptr(std::make_shared<GvlOmplPlannerHelper>(si));

    og::PathSimplifier simp(si);

    si->setStateValidityChecker(my_class_ptr->getptr());
    si->setMotionValidator(my_class_ptr->getptr());
    si->setup();


    //Create a random start state:
    ob::ScopedState<> start(space);
    start[0] = -1.3;
    start[1] = -0.2;
    start[2] = 0.0;
    start[3] = 0.0;
    start[4] = 0.0;
    start[5] = 0.0;
    //And a random goal state:
    ob::ScopedState<> goal(space);
    goal[0] = 1.3;
    goal[1] = -0.5;
    goal[2] = 0.0;
    goal[3] = 0.0;
    goal[4] = 0.0;
    goal[5] = 0.0;

    my_class_ptr->insertStartAndGoal(start, goal);
    my_class_ptr->doVis();


    //Create an instance of ompl::base::ProblemDefinition
    auto pdef(std::make_shared<ob::ProblemDefinition>(si));
    //Set the start and goal states for the problem definition.
    pdef->setStartAndGoalStates(start, goal);
    //Create an instance of a planner
    auto planner(std::make_shared<og::LBKPIECE1>(si));
    //Tell the planner which problem we are interested in solving
    planner->setProblemDefinition(pdef);
    //Make sure all the settings for the space and planner are in order. This will also lead to the runtime computation of the state validity checking resolution.
    planner->setup();

    int succs = 0;

    //use this to stop planning, until Visualizer is connected
    std::cout << "Waiting for Viz. Press Key if ready!" << std::endl;
    std::cin.ignore();

    for(size_t n = 0; n < 5; ++n)
    {
        my_class_ptr->moveObstacle();

        //We can now try to solve the problem. This call returns a value from ompl::base::PlannerStatus which describes whether a solution has been found within the specified amount of time (in seconds). If this value can be cast to true, a solution was found.
        PERF_MON_START("planner");
        planner->clear(); // this clears all roadmaps
        ob::PlannerStatus solved = planner->ob::Planner::solve(20.0);
        PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("planner", "Planning time", "planning");

        //If a solution has been found, we simplify and display it.
        if (solved)
        {
            ++succs;
            // get the goal representation from the problem definition (not the same as the goal state)
            // and inquire about the found path
            ob::PathPtr path = pdef->getSolutionPath();
            std::cout << "Found solution:" << std::endl;
            // print the path to screen
            path->print(std::cout);

            PERF_MON_START("simplify");
            simp.simplifyMax(*(path->as<og::PathGeometric>()));
            PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("simplify", "Simplification time", "planning");

            std::cout << "Simplified solution:" << std::endl;
            // print the path to screen
            path->print(std::cout);

            my_class_ptr->visualizeSolution(path);
            usleep(500000);

        }else{
            std::cout << "No solution could be found" << std::endl;
        }

        PERF_MON_SUMMARY_PREFIX_INFO("planning");

    }
    PERF_MON_ADD_STATIC_DATA_P("Number of Planning Successes", succs, "planning");

    PERF_MON_SUMMARY_PREFIX_INFO("planning");

    // keep the visualization running:
    while(true)
    {
        my_class_ptr->doVis();
        usleep(30000);
    }

    return 1;
}


gvl_ompl_planning/gvl_ompl_planner_helper.h
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK ----------------------------------------------
// This file is part of the GPU Voxels Software Library.
//
// This program is free software licensed under the CDDL
// (COMMON DEVELOPMENT AND DISTRIBUTION LICENSE Version 1.0).
// You can find a copy of this license in LICENSE.txt in the top
// directory of the source code.
//
// © Copyright 2018 FZI Forschungszentrum Informatik, Karlsruhe, Germany
//
// -- END LICENSE BLOCK ------------------------------------------------

//----------------------------------------------------------------------
/*!\file
 *
 * \author  Andreas Hermann
 * \date    2018-01-07
 *
 */
//----------------------------------------------------------------------/*
#ifndef GVL_LINKAGE_TEST_LIB_H_INCLUDED
#define GVL_LINKAGE_TEST_LIB_H_INCLUDED

#include <gpu_voxels/GpuVoxels.h>

#include <ompl/base/SpaceInformation.h>
#include <ompl/base/spaces/SE3StateSpace.h>
#include <ompl/geometric/SimpleSetup.h>

#include <ompl/config.h>
#include <iostream>
#include <tuple>

#include <mutex>


class GvlOmplPlannerHelper : public ompl::base::StateValidityChecker, public ompl::base::MotionValidator, public std::enable_shared_from_this<GvlOmplPlannerHelper>
{
public:
    GvlOmplPlannerHelper(const ompl::base::SpaceInformationPtr &si);
    ~GvlOmplPlannerHelper();

    virtual bool isValid(const ompl::base::State *state) const;
    virtual bool checkMotion(const ompl::base::State *s1, const ompl::base::State *s2,
                             std::pair< ompl::base::State*, double > & lastValid) const;
    virtual bool checkMotion(const ompl::base::State *s1, const ompl::base::State *s2) const;

    void plan();

    std::shared_ptr<GvlOmplPlannerHelper> getptr() {
        return shared_from_this();
    }

    void insertStartAndGoal(const ompl::base::ScopedState<> &start, const ompl::base::ScopedState<> &goal) const;
    void visualizeSolution(ompl::base::PathPtr path);
    void doVis();
    void moveObstacle();

private:

    gpu_voxels::GpuVoxelsSharedPtr gvl;

    ompl::base::StateSpace *stateSpace_;
    ompl::base::SpaceInformationPtr si_;

    mutable std::mutex g_i_mutex;
    mutable std::mutex g_j_mutex;
};



#endif


example_how_to_link/gvl_linkage_test_lib.cpp
#include "gvl_linkage_test_lib.h"


gvlLinkageTest::gvlLinkageTest()
{
    gvl = gpu_voxels::GpuVoxels::getInstance();
    gvl->initialize(150, 150, 150, 0.01);

    // We add maps with objects, to collide them
    gvl->addMap(gpu_voxels::MT_PROBAB_VOXELMAP,"myFirstMap");
    gvl->addMap(gpu_voxels::MT_PROBAB_VOXELMAP,"mySecondMap");

}

gvlLinkageTest::~gvlLinkageTest()
{
    gvl.reset(); // Not even required, as we use smart pointers.
}

int gvlLinkageTest::doFancyStuff()
{
    gpu_voxels::Vector3f center = gpu_voxels::Vector3f(0.5 , 0.5, 0.5);
    gpu_voxels::Vector3f box_size = gpu_voxels::Vector3f(0.4 , 0.4, 0.4);
    gpu_voxels::Vector3f overlap = gpu_voxels::Vector3f(0.1 , 0.1, 0.1);

    gpu_voxels::Vector3f c1 = center - overlap;
    gpu_voxels::Vector3f c2 = c1 + box_size;

    gpu_voxels::Vector3f c3 = center + overlap;
    gpu_voxels::Vector3f c4 = c3 - box_size;

    gvl->insertBoxIntoMap(c1, c2, "myFirstMap", gpu_voxels::eBVM_OCCUPIED, 1);
    gvl->insertBoxIntoMap(c4, c3, "mySecondMap", gpu_voxels::eBVM_OCCUPIED, 1);

    size_t num_colls_pc = gvl->getMap("myFirstMap")->as<gpu_voxels::voxelmap::ProbVoxelMap>()->collideWith(gvl->getMap("mySecondMap")->as<gpu_voxels::voxelmap::ProbVoxelMap>());
    LOGGING_INFO(gpu_voxels::Gpu_voxels, num_colls_pc << " collisions detected" << gpu_voxels::endl );

    return 0;
}


gvl_ompl_planning/gvl_ompl_planner_helper.cpp
// this is for emacs file handling -*- mode: c++; indent-tabs-mode: nil -*-

// -- BEGIN LICENSE BLOCK ----------------------------------------------
// This file is part of the GPU Voxels Software Library.
//
// This program is free software licensed under the CDDL
// (COMMON DEVELOPMENT AND DISTRIBUTION LICENSE Version 1.0).
// You can find a copy of this license in LICENSE.txt in the top
// directory of the source code.
//
// © Copyright 2018 FZI Forschungszentrum Informatik, Karlsruhe, Germany
//
// -- END LICENSE BLOCK ------------------------------------------------

//----------------------------------------------------------------------
/*!\file
 *
 * \author  Andreas Hermann
 * \date    2018-01-07
 *
 */
//----------------------------------------------------------------------/*
#include "gvl_ompl_planner_helper.h"

#include <boost/filesystem/operations.hpp>
#include <boost/filesystem/path.hpp>

#include <gpu_voxels/GpuVoxels.h>
#include <gpu_voxels/helpers/MetaPointCloud.h>
#include <gpu_voxels/robot/urdf_robot/urdf_robot.h>
#include <gpu_voxels/logging/logging_gpu_voxels.h>
#include <thread>

#define IC_PERFORMANCE_MONITOR
#include <icl_core_performance_monitor/PerformanceMonitor.h>

namespace ob = ompl::base;
namespace og = ompl::geometric;
using namespace gpu_voxels;
namespace bfs = boost::filesystem;

GvlOmplPlannerHelper::GvlOmplPlannerHelper(const ob::SpaceInformationPtr &si)
    : ob::StateValidityChecker(si)
    , ob::MotionValidator(si)
{

    si_ = si;
    stateSpace_ = si_->getStateSpace().get();
    assert(stateSpace_ != nullptr);

    gvl = gpu_voxels::GpuVoxels::getInstance();
    gvl->initialize(150, 150, 100, 0.02);

    // We add maps with objects, to collide them
    gvl->addMap(MT_PROBAB_VOXELMAP,"myRobotMap");
    gvl->addMap(MT_PROBAB_VOXELMAP,"myEnvironmentMap");
    gvl->addMap(MT_BITVECTOR_VOXELLIST,"mySolutionMap");
    gvl->addMap(MT_PROBAB_VOXELMAP,"myQueryMap");

    gvl->addRobot("myUrdfRobot", "ur10_coarse/ur10_joint_limited_robot.urdf", true);

    gvl->visualizeMap("myEnvironmentMap");

    PERF_MON_ENABLE("pose_check");
    PERF_MON_ENABLE("motion_check");
    PERF_MON_ENABLE("motion_check_lv");
}



GvlOmplPlannerHelper::~GvlOmplPlannerHelper()
{
    gvl.reset(); // Not even required, as we use smart pointers.
}

void GvlOmplPlannerHelper::moveObstacle()
{
    gvl->clearMap("myEnvironmentMap");
    static float x(1.0);

    // use this to animate a single moving box obstacle
    //gvl->insertBoxIntoMap(Vector3f(2.0, x ,0.0), Vector3f(2.2, x + 0.2 ,1.2), "myEnvironmentMap", eBVM_OCCUPIED, 2);
    x += 0.1;

    gvl->insertBoxIntoMap(Vector3f(1.0,1.0,0.0), Vector3f(1.2,1.2,1.2), "myEnvironmentMap", eBVM_OCCUPIED, 2);
    gvl->insertBoxIntoMap(Vector3f(1.8,1.8,0.0), Vector3f(2.0,2.0,1.2), "myEnvironmentMap", eBVM_OCCUPIED, 2);
    gvl->insertBoxIntoMap(Vector3f(1.1,1.1,1.2), Vector3f(1.9,1.9,1.3), "myEnvironmentMap", eBVM_OCCUPIED, 2);
    gvl->insertBoxIntoMap(Vector3f(0.0,0.0,0.0), Vector3f(3.0,3.0,0.01), "myEnvironmentMap", eBVM_OCCUPIED, 2);
    gvl->visualizeMap("myEnvironmentMap");
}

void GvlOmplPlannerHelper::doVis()
{

    // tell the visualier that the map has changed:
    gvl->visualizeMap("myRobotMap");
    gvl->visualizeMap("myEnvironmentMap");
    gvl->visualizeMap("mySolutionMap");
    gvl->visualizeMap("myQueryMap");
}

void GvlOmplPlannerHelper::visualizeSolution(ob::PathPtr path)
{
    gvl->clearMap("mySolutionMap");

    PERF_MON_SUMMARY_PREFIX_INFO("pose_check");
    PERF_MON_SUMMARY_PREFIX_INFO("motion_check");
    PERF_MON_SUMMARY_PREFIX_INFO("motion_check_lv");

    std::cout << "Robot consists of " << gvl->getRobot("myUrdfRobot")->getTransformedClouds()->getAccumulatedPointcloudSize() << " points" << std::endl;

    og::PathGeometric* solution = path->as<og::PathGeometric>();
    solution->interpolate();


    for(size_t step = 0; step < solution->getStateCount(); ++step)
    {

        const double *values = solution->getState(step)->as<ob::RealVectorStateSpace::StateType>()->values;

        robot::JointValueMap state_joint_values;
        state_joint_values["shoulder_pan_joint"] = values[0];
        state_joint_values["shoulder_lift_joint"] = values[1];
        state_joint_values["elbow_joint"] = values[2];
        state_joint_values["wrist_1_joint"] = values[3];
        state_joint_values["wrist_2_joint"] = values[4];
        state_joint_values["wrist_3_joint"] = values[5];

        // update the robot joints:
        gvl->setRobotConfiguration("myUrdfRobot", state_joint_values);
        // insert the robot into the map:
        gvl->insertRobotIntoMap("myUrdfRobot", "mySolutionMap", BitVoxelMeaning(eBVM_SWEPT_VOLUME_START + (step % 249) ));
    }

    gvl->visualizeMap("mySolutionMap");

}

void GvlOmplPlannerHelper::insertStartAndGoal(const ompl::base::ScopedState<> &start, const ompl::base::ScopedState<> &goal) const
{

    gvl->clearMap("myQueryMap");

    robot::JointValueMap state_joint_values;
    state_joint_values["shoulder_pan_joint"] = start[0];
    state_joint_values["shoulder_lift_joint"] = start[1];
    state_joint_values["elbow_joint"] = start[2];
    state_joint_values["wrist_1_joint"] = start[3];
    state_joint_values["wrist_2_joint"] = start[4];
    state_joint_values["wrist_3_joint"] = start[5];

    // update the robot joints:
    gvl->setRobotConfiguration("myUrdfRobot", state_joint_values);
    gvl->insertRobotIntoMap("myUrdfRobot", "myQueryMap", BitVoxelMeaning(eBVM_SWEPT_VOLUME_START));

    state_joint_values["shoulder_pan_joint"] = goal[0];
    state_joint_values["shoulder_lift_joint"] = goal[1];
    state_joint_values["elbow_joint"] = goal[2];
    state_joint_values["wrist_1_joint"] = goal[3];
    state_joint_values["wrist_2_joint"] = goal[4];
    state_joint_values["wrist_3_joint"] = goal[5];

    // update the robot joints:
    gvl->setRobotConfiguration("myUrdfRobot", state_joint_values);
    gvl->insertRobotIntoMap("myUrdfRobot", "myQueryMap", BitVoxelMeaning(eBVM_SWEPT_VOLUME_START+1));

}

bool GvlOmplPlannerHelper::isValid(const ompl::base::State *state) const
{

    PERF_MON_START("inserting");

    std::lock_guard<std::mutex> lock(g_i_mutex);

    gvl->clearMap("myRobotMap");

    const double *values = state->as<ob::RealVectorStateSpace::StateType>()->values;

    robot::JointValueMap state_joint_values;
    state_joint_values["shoulder_pan_joint"] = values[0];
    state_joint_values["shoulder_lift_joint"] = values[1];
    state_joint_values["elbow_joint"] = values[2];
    state_joint_values["wrist_1_joint"] = values[3];
    state_joint_values["wrist_2_joint"] = values[4];
    state_joint_values["wrist_3_joint"] = values[5];

    // update the robot joints:
    gvl->setRobotConfiguration("myUrdfRobot", state_joint_values);
    // insert the robot into the map:
    gvl->insertRobotIntoMap("myUrdfRobot", "myRobotMap", eBVM_OCCUPIED);

    PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("insert", "Pose Insertion", "pose_check");

    PERF_MON_START("coll_test");
    size_t num_colls_pc = gvl->getMap("myRobotMap")->as<voxelmap::ProbVoxelMap>()->collideWith(gvl->getMap("myEnvironmentMap")->as<voxelmap::ProbVoxelMap>(), 0.7f);
    PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("coll_test", "Pose Collsion", "pose_check");

    //std::cout << "Validity check on state ["  << values[0] << ", " << values[1] << ", " << values[2] << ", " << values[3] << ", " << values[4] << ", " << values[5] << "] resulting in " <<  num_colls_pc << " colls." << std::endl;

    return num_colls_pc == 0;
}

bool GvlOmplPlannerHelper::checkMotion(const ompl::base::State *s1, const ompl::base::State *s2,
                                       std::pair< ompl::base::State*, double > & lastValid) const
{

    //    std::cout << "LongestValidSegmentFraction = " << stateSpace_->getLongestValidSegmentFraction() << std::endl;
    //    std::cout << "getLongestValidSegmentLength = " << stateSpace_->getLongestValidSegmentLength() << std::endl;
    //    std::cout << "getMaximumExtent = " << stateSpace_->getMaximumExtent() << std::endl;


    std::lock_guard<std::mutex> lock(g_j_mutex);
    gvl->clearMap("myRobotMap");

    /* assume motion starts in a valid configuration so s1 is valid */

    bool result = true;
    int nd = stateSpace_->validSegmentCount(s1, s2);

    //std::cout << "Called interpolating motion_check_lv to evaluate " << nd << " segments" << std::endl;

    PERF_MON_ADD_DATA_NONTIME_P("Num poses in motion", float(nd), "motion_check_lv");
    if (nd > 1)
    {
        /* temporary storage for the checked state */
        ob::State *test = si_->allocState();

        for (int j = 1; j < nd; ++j)
        {
            stateSpace_->interpolate(s1, s2, (double)j / (double)nd, test);
            if (!si_->isValid(test))

            {
                lastValid.second = (double)(j - 1) / (double)nd;
                if (lastValid.first != nullptr)
                    stateSpace_->interpolate(s1, s2, lastValid.second, lastValid.first);
                result = false;
                break;

            }
        }

        si_->freeState(test);

    }

    if (result)
        if (!si_->isValid(s2))
        {
            lastValid.second = (double)(nd - 1) / (double)nd;
            if (lastValid.first != nullptr)
                stateSpace_->interpolate(s1, s2, lastValid.second, lastValid.first);
            result = false;
        }


    if (result)
        valid_++;
    else
        invalid_++;


    return result;
}

bool GvlOmplPlannerHelper::checkMotion(const ompl::base::State *s1, const ompl::base::State *s2) const
{
    std::lock_guard<std::mutex> lock(g_i_mutex);
    gvl->clearMap("myRobotMap");


    //        std::cout << "LongestValidSegmentFraction = " << stateSpace_->getLongestValidSegmentFraction() << std::endl;
    //        std::cout << "getLongestValidSegmentLength = " << stateSpace_->getLongestValidSegmentLength() << std::endl;
    //        std::cout << "getMaximumExtent = " << stateSpace_->getMaximumExtent() << std::endl;



    /* assume motion starts in a valid configuration so s1 is valid */

    bool result = true;
    int nd = stateSpace_->validSegmentCount(s1, s2);

    // not required with ProbabVoxels:
    //    if(nd > 249)
    //    {
    //        std::cout << "Too many intermediate states for BitVoxels" << std::endl;
    //        exit(1);
    //    }

    if (nd > 1)
    {
        PERF_MON_START("inserting");

        /* temporary storage for the checked state */
        ob::State *test = si_->allocState();

        for (int j = 1; j < nd; ++j)
        {
            stateSpace_->interpolate(s1, s2, (double)j / (double)nd, test);


            const double *values = test->as<ob::RealVectorStateSpace::StateType>()->values;

            robot::JointValueMap state_joint_values;
            state_joint_values["shoulder_pan_joint"] = values[0];
            state_joint_values["shoulder_lift_joint"] = values[1];
            state_joint_values["elbow_joint"] = values[2];
            state_joint_values["wrist_1_joint"] = values[3];
            state_joint_values["wrist_2_joint"] = values[4];
            state_joint_values["wrist_3_joint"] = values[5];

            // update the robot joints:
            gvl->setRobotConfiguration("myUrdfRobot", state_joint_values);
            // insert the robot into the map:
            gvl->insertRobotIntoMap("myUrdfRobot", "myRobotMap", eBVM_OCCUPIED);

        }
        PERF_MON_ADD_DATA_NONTIME_P("Num poses in motion", float(nd), "motion_check");

        si_->freeState(test);

        PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("insert", "Motion Insertion", "motion_check");

        //gvl->visualizeMap("myRobotMap");
        PERF_MON_START("coll_test");
        size_t num_colls_pc = gvl->getMap("myRobotMap")->as<voxelmap::ProbVoxelMap>()->collideWith(gvl->getMap("myEnvironmentMap")->as<voxelmap::ProbVoxelMap>(), 0.7f);
        //std::cout << "CheckMotion1 for " << nd << " segments. Resulting in " << num_colls_pc << " colls." << std::endl;
        PERF_MON_SILENT_MEASURE_AND_RESET_INFO_P("coll_test", "Pose Collsion", "motion_check");

        result = (num_colls_pc == 0);

    }


    if (result)
        valid_++;
    else
        invalid_++;


    return result;


}
